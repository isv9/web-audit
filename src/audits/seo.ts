import { AuditResult, AuditResultLog, WebDocument } from '../web-audit';

type BaseOpenGraphMetaPropertyName = 'title' | 'description' | 'url';
type OpenGraphMetaPropertyName = BaseOpenGraphMetaPropertyName;
type CheckedOpenGraphMetaTag = { propertyName: OpenGraphMetaPropertyName; isExisted: boolean };

export function auditOpenGraph(document: Pick<WebDocument, 'querySelectorAll'>): AuditResult {
  const baseOgMetaPropertiesNames: BaseOpenGraphMetaPropertyName[] = ['title', 'description', 'url'];
  const checkedBaseOpenGraphMetaTags: CheckedOpenGraphMetaTag[] = baseOgMetaPropertiesNames.map(
    (propertyName): CheckedOpenGraphMetaTag => ({
      propertyName,
      isExisted: checkExistOpenGraphMetaTag(document, propertyName),
    }),
  );
  const warnings: string[] = [];
  if (checkedBaseOpenGraphMetaTags.some(({ isExisted }) => !isExisted)) {
    checkedBaseOpenGraphMetaTags.forEach(({ isExisted, propertyName }) => {
      if (!isExisted) {
        warnings.push(`Document does not have a base meta tag 'og:${propertyName}'`);
      }
    });
  }

  const logs: AuditResultLog[] = [];
  if (checkedBaseOpenGraphMetaTags.some(({ isExisted }) => isExisted)) {
    const existedOpenGraphBaseMetaPropertiesNames = checkedBaseOpenGraphMetaTags
      .filter(({ isExisted }) => isExisted)
      .map(({ propertyName }) => propertyName);
    logs.push(`Document has some base meta tags: '${existedOpenGraphBaseMetaPropertiesNames.join(', ')}'`);
  }

  return {
    warnings,
    logs,
    name: 'Open Graph',
    tables: [],
  };
}

function checkExistOpenGraphMetaTag(
  document: Pick<WebDocument, 'querySelectorAll'>,
  ogMetaPropertyName: OpenGraphMetaPropertyName,
): boolean {
  return document.querySelectorAll(createOpenGraphMetaPropertyQuerySelector(ogMetaPropertyName)).length > 0;
}

export function createOpenGraphMetaPropertyQuerySelector(ogMetaPropertyName: OpenGraphMetaPropertyName): string {
  return `meta[property="og:${ogMetaPropertyName}"]`;
}
