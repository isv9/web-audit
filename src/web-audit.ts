import { auditBlockSemantics, auditCommonSemantics, auditHeaderSemantics, auditTextSemantics } from './audits/semantic'
import { auditLinks } from './audits/links'

export type WebDocument = {
  getElementsByTagNameCount (tag: string): number;
  querySelectorAll (query: string): { length: number };
  getEmptyElementsByTagName (tag: string): { length: number }
};

type AuditSectionResult = {
  name: string;
  auditResults: AuditResult[];
};

export type AuditResult = {
  name?: string;
  tables: (AuditResultTable)[];
  liveCollections?: (AuditResultLiveCollection)[];
  logs?: string[];
  warnings?: string[];
  errors?: string[];
};

type AuditResultTable = { name?: string } & { [key: string]: string | number };
export type AuditResultLiveCollection = { name: string, collection: object };

export class WebAudit {

  private readonly webDocument: WebDocument

  constructor (webDocument: WebDocument) {
    this.webDocument = webDocument
  }

  audit () {
    const auditSemanticsResult = this.auditSemantics()
    const auditLinksResult = this.auditLinks()
    console.group('web audit')
    WebAudit.renderAuditSectionResult(auditSemanticsResult)
    WebAudit.renderAuditSectionResult(auditLinksResult)
    console.groupEnd()
  }

  private auditLinks (): AuditSectionResult {
    return {
      name: 'links',
      auditResults: [
        auditLinks(this.webDocument)]
    }
  }

  private auditSemantics (): AuditSectionResult {

    return {
      name: 'semantics',
      auditResults: [auditCommonSemantics(this.webDocument),
        auditHeaderSemantics(this.webDocument),
        auditTextSemantics(this.webDocument),
        auditBlockSemantics(this.webDocument)]
    }
  }

  static renderAuditSectionResult (auditSectionResult: AuditSectionResult) {
    const { name, auditResults } = auditSectionResult
    const details = {
      errorsCount: 0,
      warningsCount: 0
    }
    auditResults.forEach(auditResult => {
      const errorsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'errors')
      const warningsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'warnings')
      details.errorsCount += errorsCount
      details.warningsCount += warningsCount
    })
    console.group(`${name}, (errors="${details.errorsCount}", warnings="${details.warningsCount}")`)
    auditResults.forEach(auditResult => WebAudit.renderAuditResult(auditResult))
    console.groupEnd()
  }

  static renderAuditResult (auditResult: AuditResult) {
    const { name, tables, errors = [], warnings = [], liveCollections = [], logs = [] } = auditResult
    const errorsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'errors')
    const warningsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'warnings')
    console.group(`${name}, (errors="${errorsCount}", warnings="${warningsCount}")`)
    tables.forEach(({ name, nodes, ...tableDataSet }) => {
      console.log(name)
      console.table(tableDataSet)
    })
    liveCollections.forEach(({ name, collection }) => {
      console.log(name)
      console.log(collection)
    })
    logs.forEach(message => console.log(message))
    console.log('Summary')
    if (errorsCount > 0 || warningsCount > 0) {
      errors.forEach(message => console.error(message))
      warnings.forEach(message => console.warn(message))
    } else {
      console.log('It is OK')
    }

    console.groupEnd()
  }

  static getAuditResultMessagesCount (auditResult: AuditResult, key: keyof AuditResult): number {
    let property = auditResult[key]
    if (property === null || property === undefined) {
      property = []
    }
    if (Array.isArray(property)) {
      return property.length
    }
    throw new Error('Property type must be array')
  }
}

