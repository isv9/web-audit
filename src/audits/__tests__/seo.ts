import { WebDocument } from '../../web-audit';
import { auditOpenGraph, createOpenGraphMetaPropertyQuerySelector } from '../seo';

describe('auditOpenGraph', () => {
  it('document has all Open Graph base tags', () => {
    const document: Pick<WebDocument, 'querySelectorAll'> = {
      querySelectorAll: jest.fn(() => ({ length: 1 })),
    };
    expect(auditOpenGraph(document)).toMatchSnapshot();
  });

  it('document has only Open Graph title tag from base tags', () => {
    const document: Pick<WebDocument, 'querySelectorAll'> = {
      querySelectorAll: jest.fn(querySelector => ({
        length: querySelector === 'meta[property="og:title"]' ? 1 : 0,
      })),
    };
    expect(auditOpenGraph(document)).toMatchSnapshot();
  });

  it('createOpenGraphMetaPropertyQuerySelector', () => {
    expect(createOpenGraphMetaPropertyQuerySelector('title')).toMatchSnapshot();
  });
});
