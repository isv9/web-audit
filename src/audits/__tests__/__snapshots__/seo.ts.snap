// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auditOpenGraph createOpenGraphMetaPropertyQuerySelector 1`] = `"meta[property=\\"og:title\\"]"`;

exports[`auditOpenGraph document has all Open Graph base tags 1`] = `
Object {
  "logs": Array [
    "Document has some base meta tags: 'title, description, url'",
  ],
  "name": "Open Graph",
  "warnings": Array [],
}
`;

exports[`auditOpenGraph document has only Open Graph title tag from base tags 1`] = `
Object {
  "logs": Array [
    "Document has some base meta tags: 'title'",
  ],
  "name": "Open Graph",
  "warnings": Array [
    "Document does not have a base meta tag 'og:description'",
    "Document does not have a base meta tag 'og:url'",
  ],
}
`;
